### DISCUSSION TITLE
Merging pretrained pose estimation with custom pose estimation in same model

### DISCUSSION THREAD

**Discussion by creeperstudio1** [ultralytics/ultralytics#13847] [Q&A]
YOLOv8 has a pretrained pose-estimation model which detects different key-points of a class: person. This dataset was trained using the coco-dataset which contains thousands of images. I have a custom dataset which contains around 600 images of another class in which I already annotated the key-points and its bounding box. This class is called robot. So, my question is, how can I train a model that detects and estimates the key-points of both the robot and the person. We are using a camera in which we would like to display both classes in the webcam or in the same frames.

**Comment by glenn-jocher** [comment]
@creeperstudio1 hello!

Thank you for reaching out with your question. It's great to see your interest in extending the YOLOv8 pose estimation model to include both human and robot keypoints.

To achieve this, you can follow these steps:

1. **Prepare Your Dataset**: Ensure that your custom dataset (robot) is in the Ultralytics YOLO format. This involves having one text file per image, with each row containing the class index, bounding box coordinates, and keypoint coordinates. For more details on the format, you can refer to the [Ultralytics documentation on pose estimation datasets](https://docs.ultralytics.com/datasets/pose/).

2. **Merge Datasets**: Combine your custom dataset with the COCO dataset. You can do this by merging the image and label files into a single directory structure. Update the YAML file to reflect the paths to your combined dataset.

3. **Update YAML Configuration**: Create or update your dataset YAML file to include both classes (person and robot). Here is an example configuration:
    ```yaml
    path: ../datasets/combined-pose  # dataset root dir
    train: images/train  # train images (relative to 'path')
    val: images/val  # val images (relative to 'path')
    test:  # test images (optional)

    # Keypoints
    kpt_shape: [17, 3]  # number of keypoints, number of dims (2 for x,y or 3 for x,y,visible)
    flip_idx: [0, 2, 1, 4, 3, 6, 5, 8, 7, 10, 9, 12, 11, 14, 13, 16, 15]

    # Classes dictionary
    names:
      0: person
      1: robot
    ```

4. **Train the Model**: Use the YOLOv8 model to train on your combined dataset. Here is an example of how to do this in Python:
    ```python
    from ultralytics import YOLO

    # Load a pretrained model
    model = YOLO("yolov8n-pose.pt")

    # Train the model on the combined dataset
    results = model.train(data="combined-pose.yaml", epochs=100, imgsz=640)
    ```

    Or via the CLI:
    ```bash
    yolo detect train data=combined-pose.yaml model=yolov8n-pose.pt epochs=100 imgsz=640
    ```

5. **Validate and Test**: After training, validate the model to ensure it performs well on both classes. You can use the `val` mode to evaluate the model on your validation set.

By following these steps, you should be able to train a model that can detect and estimate keypoints for both humans and robots in the same frame. If you encounter any issues or have further questions, feel free to ask!

Best of luck with your project! 🚀
